src/Compilers/CSharp/Test/Semantic/Semantics/StructConstructorTests.cs

```diff
@@ -0,0 +1,1329 @@
+{
+    static void Main()
+    {
+        Console.WriteLine(new S1(null).Initialized);
+        Console.WriteLine(new S2().Initialized);
+    }
+}";
+
+            var verifier = CompileAndVerify(source, parseOptions: TestOptions.RegularPreview, expectedOutput:
+@"True
+True");
+            verifier.VerifyIL("S1..ctor()",
+@"{
+  // Code size        8 (0x8)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  ldc.i4.1
+  IL_0002:  stfld      ""bool S1.Initialized""
+  IL_0007:  ret
+}");
+            verifier.VerifyIL("S1..ctor(object)",
+@"{
+  // Code size        7 (0x7)
+  .maxstack  1
+  IL_0000:  ldarg.0
+  IL_0001:  call       ""S1..ctor()""
+  IL_0006:  ret
+}");
+            verifier.VerifyIL("S2..ctor()",
+@"{
+  // Code size        8 (0x8)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  ldnull
+  IL_0002:  call       ""S2..ctor(object)""
+  IL_0007:  ret
+}");
+            verifier.VerifyIL("S2..ctor(object)",
+@"{
+  // Code size        8 (0x8)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  ldc.i4.1
+  IL_0002:  stfld      ""bool S2.Initialized""
+  IL_0007:  ret
+}");
+        }
+
+        [Fact]
+        public void ThisInitializer_02()
+        {
+            var source =
+@"using System;
+struct S1
+{
+    internal bool Initialized;
+    public S1() { Initialized = false; }
+    internal S1(object obj) : this() { Initialized = true; }
+}
+struct S2
+{
+    internal bool Initialized;
+    public S2() : this(null) { Initialized = true; }
+    S2(object obj) { Initialized = false; }
+}
+class Program
+{
+    static void Main()
+    {
+        Console.WriteLine(new S1(null).Initialized);
+        Console.WriteLine(new S2().Initialized);
+    }
+}";
+
+            var verifier = CompileAndVerify(source, parseOptions: TestOptions.RegularPreview, expectedOutput:
+@"True
+True");
+            verifier.VerifyIL("S1..ctor()",
+@"{
+  // Code size        8 (0x8)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  ldc.i4.0
+  IL_0002:  stfld      ""bool S1.Initialized""
+  IL_0007:  ret
+}");
+            verifier.VerifyIL("S1..ctor(object)",
+@"{
+  // Code size       14 (0xe)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  call       ""S1..ctor()""
+  IL_0006:  ldarg.0
+  IL_0007:  ldc.i4.1
+  IL_0008:  stfld      ""bool S1.Initialized""
+  IL_000d:  ret
+}");
+            verifier.VerifyIL("S2..ctor()",
+@"{
+  // Code size       15 (0xf)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  ldnull
+  IL_0002:  call       ""S2..ctor(object)""
+  IL_0007:  ldarg.0
+  IL_0008:  ldc.i4.1
+  IL_0009:  stfld      ""bool S2.Initialized""
+  IL_000e:  ret
+}");
+            verifier.VerifyIL("S2..ctor(object)",
+@"{
+  // Code size        8 (0x8)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  ldc.i4.0
+  IL_0002:  stfld      ""bool S2.Initialized""
+  IL_0007:  ret
+}");
+        }
+
+        [Fact]
+        public void ThisInitializer_03()
+        {
+            var source =
+@"using System;
+struct S0
+{
+    internal bool Initialized = true;
```

Comment:
nit: I think the testing approach is okay, because the IL is fairly simple, but consider using printing methods instead to show evaluation order in expectedOutput. `bool Initialized = Print(true);` and such.